xc/device.ml

```diff
@@ -1444,22 +1461,34 @@ module Vusb = struct
       end
     in
     let is_running = Qemu.is_running ~xs domid in
-    match is_running with
-    | true  ->
-      let cmd = Qmp.(Device_add
-                       ( "usb-host"
-                       , id
-                       , Some (get_bus version, hostbus, hostport)
-                       ))
-      in qmp_send_cmd domid cmd |> ignore
-    | _ -> ()
-
-  let vusb_unplug ~xs ~domid ~id =
+    if is_running then
+      begin
+        (* Need to reset USB device before passthrough to vm according to CP-24616.
+           Also need to do deprivileged work in usb_reset script if QEMU is deprivileged.
+        *)
+        begin match Qemu.pid ~xs domid with
+          | Some pid -> call_usb_reset hostbus hostport domid pid "attach" priv
+          | _ -> failwith (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid))
+        end;
+        let cmd = Qmp.(Device_add
+                         ( "usb-host"
+                         , id
+                         , Some (get_bus version, hostbus, hostport)
+                         ))
+        in qmp_send_cmd domid cmd |> ignore
+      end
+
+  let vusb_unplug ~xs ~priv ~domid ~id ~hostbus ~hostport=
     debug "Vusb unplugged: vusb device unplugged";
     let is_running = Qemu.is_running ~xs domid in
-    match is_running with
-    | true  -> qmp_send_cmd domid Qmp.(Device_del id) |> ignore
-    | _     -> ()
+    if is_running then
+      begin
+        qmp_send_cmd domid Qmp.(Device_del id) |> ignore;
```

Comment:
Hi @lindig , if qmp_send_cmd is failed, seems the error will be raised and the vusb cleanup below won't be called. I have tested that when a vm is shut down with usb attached, I will catch an error "Caught (Unix.Unix_error "Connection refused" connect "")". IMO, when a vm is shut down, it calls vusb_unplug but the qmp socket is closed which raise a connection error. Am I right?

Reply:
If qmp_send fails, it will raise an error. If you need to call the cleanup action in any case, this should be wrapped in a `finally` call.

Reply:
If I unplug vusb firstly and then shut down vm, will not get the error.  What I mean here is that the uid/gid of the device file is not restored when catching the connection error, as the call_usb_reset is not called.

Reply:
OK. But we still can get the error in XenCenter. Is this ok?

Reply:
XenCenter receives its errors from xapi, which receives errors from xenopsd. If this is an unexpected error, this should be reported as `Internal_error` which will propagate to XenCenter. However, this should only be raised (or re-raised) after all cleanup actions have been performed.

Reply:
Also, if this error needs to have a proper error message in XenCenter, you need to define a new error in the idl, raise that and deal with it in xapi.